gh-118164: str(10**10000) hangs if the C _decimal module is missing

tim-one opened the pull request:
This is somewhat of a conceptual mess, related to two issues:

https://github.com/python/cpython/issues/118164
> str(10**10000) hangs if the C _decimal module is missing

https://github.com/python/cpython/issues/118027
> _pydecimal.Decimal.__pow__ tries to compute 10**1000000000000000000 and thus doesn't terminate

These only occur if the C `decimal` module isn't available, so `_pydecimal.py` is used instead. But CPython doesn't use `_pydecimal.py` for anything anymore, so this can only happen in a damaged CPython installation.. Or, significantly, under PyPy, which hasn't incorporated the C `libmpdec` CPython uses to implement `decimal`.

Confounding it is that the first issue above can't be tested even after _this_ PR's changes. This PR repairs the proximate cause of the failure, which causes it to fail instead later.  That failure is the subject of a different PR:

https://github.com/python/cpython/pull/118483

If someone can think of a sane way to test this, I'd be grateful :smile:l.




<!-- gh-issue-number: gh-118164 -->
* Issue: gh-118164
<!-- /gh-issue-number -->

<!-- gh-issue-number: gh-118027 -->
* Issue: gh-118027
<!-- /gh-issue-number -->

tim-one committed:
Initial stab.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into pydec2

pochmann3 commented on Lib/_pydecimal.py:
```diff
@@ -2131,10 +2164,11 @@ def _power_exact(self, other, p):
 
-            if xc >= 10**p:
+            strxc = _convert_to_str(xc, p)
+            if strxc is None:
```
Is this equivalent in context? `xc >= 10**p` and `_convert_to_str(xc, p) is None` aren't equivalent for example for `xc = 1000000` and `p = 3`, as `_convert_to_str` only checks for *any* power of 10, not for just the p-th power of 10. Are larger powers of 10 ruled out some other way?

tim-one replied:
I believe larger powers are already ruled out, but am not 100% sure of that - the code is very involved.

I'm relying in the end on the stated _purpose_ of the "check exact" context, which is to determine whether the input is exactly representable in decimal floating point with `p` digits of precision. `10**i` for any `i >= 0` is, provided only `p >= 1`. I'm confident enough in that to skip adding a check that the string has exactly `p+1` digits too.

tim-one replied:
I added a new comment to spell this out. The new behavior may or may not be right, and the old behavior may or may not be wrong. Had to take my best guess, because I've had little luck so far trying to out-think this code ☹️

tim-one committed:
Merge remote-tracking branch 'upstream/main' into pydec2

tim-one committed:
Test the tentative fix. Hangs "forever" without this change.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into pydec2

tim-one committed:
Move the new test to a better spot.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into pydec2

tim-one committed:
Merge remote-tracking branch 'upstream/main' into pydec2

tim-one committed:
New comment to explain why _convert_to_str allows any poewr of 10.

tim-one committed:
Fixed a comment, and fleshed out an existing test that appeared unfinished.

tim-one committed:
Merge branch 'main' into pydec2

tim-one committed:
Added temporary asserts. Or maybe permanent ;-)

serhiy-storchaka commented on Lib/_pydecimal.py:
```diff
@@ -2131,10 +2181,11 @@ def _power_exact(self, other, p):
             else:
                 return None
 
-            if xc >= 10**p:
```
I have read the code above, it is not so complex, and I can guarantee that `xc` is never the power of 10.

From the start, it is made non-divisible by 10. The case `xc == 1` is handled separately. Then the only transformation of `xc` is taking the `n`th root (exact) and elevation to the `m`th degree.

tim-one replied:
I agree - and already said the same (with different words) in comments on the PR :smile:.

In the other place the test is made, it's even easier to show that power-of-10 can't happen.

Thanks for looking at it!

serhiy-storchaka commented on Lib/_pydecimal.py:
```diff
@@ -2180,17 +2231,19 @@ def _power_exact(self, other, p):
         # if m > p*100//_log10_lb(xc) then m > p/log10(xc), hence xc**m >
         # 10**p and the result is not representable.
+        assert xc != 1, self
+        assert not _is_power_of_10(str(xc)), self
```
You can use more strong and cheaper assertion:
```suggestion
        assert xc % 10 != 0, self
```

tim-one committed:
Update Lib/_pydecimal.py

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

tim-one committed:
Merge remote-tracking branch 'upstream/main' into pydec2

tim-one committed:
Remove the new _convert_to_str().

Serhiy and I independently concluded that exact powers of 10
aren't possible in these contexts, so just checking the
string length is sufficient.

tim-one committed:
At least for now, add the asserts to the other block too.

serhiy-storchaka reviewed (approved):
LGTM.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.
